python/cudf/cudf/core/dtypes.py

```diff
@@ -31,6 +33,10 @@ def type(self):
     def name(self):
         return "category"
 
+    @property
+    def kind(self):
+        return np.dtype(min_signed_type(len(self._categories))).kind
```

Comment:
This seems like a bad assumption, can we just check the child dtype of the column as opposed to it being baked into the CategoricalDtype?

Reply:
So this is here for two reasons.

First, `CategoricalDtype.kind` will report the kind 'O' without this definition, even if the categories aren't strings. This behavior trips up checks like this: https://github.com/rapidsai/cudf/blob/73e1db07ab5dfb250b24474fa194936769dd0ecf/python/cudf/cudf/_libxx/transpose.pyx#L30-L31

But before I assumed anything, I checked with pandas. Here's what the [`pd.ExtensionDtype.kind` docstring](https://github.com/pandas-dev/pandas/blob/ac3056f2f13e5287dc66d09f31522f7fba592fec/pandas/core/dtypes/base.py#L156-L161) says `kind` should report:
> A character code (one of 'biufcmMOSUV'), default 'O'
> 
> This should match the NumPy dtype used when the array is converted to an ndarray, which is probably 'O' for object if the extension type cannot be represented as a built-in NumPy type.

So I checked the dtype of the numpy array, and sure enough:
```python
assert categorical_sr.to_array().dtype.kind == 'i' # True
```

I'd rather return `codes.dtype.kind` here, but can't since `CategoricalDtype` only has access to the _categories_.

Reply:
I think the fact that we return the codes when calling `.to_array()` is actually a bug as opposed to intended behavior. We should probably materialize the column in this situation and return an array based on the materialization.